### PR TITLE
Support reconnection in puppetjs. Work in progress

### PR DESCRIPTION
**Pull request by joozek78**

### CODE REVIEW

File: src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs [hunk 1]
```diff
@@ -101,8 +103,16 @@ internal static class PuppetRestHandler {
...
+                            Resource = root,
                             StatusCode = 202,
-                            StatusDescription = "Patch enqueued until earlier versions have arrived"
+                            StatusDescription = "Patch enqueued until earlier versions have arrived. Last known version is " + root.ChangeLog.Version.RemoteVersion
+                        };
+                    } else if (patchCount == -2) {
+                        return new Response()
+                        {
+                            Resource = root,
+                            StatusCode = 203,
```
Comment (joozek78) [inline src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs]: I'm wondering whether we should differentiate this case, or maybe just return normally

File: src/Starcounter.XSON/Sessions/JSON-Patch/JsonPatch.cs [hunk 2]
```diff
@@ -655,7 +656,8 @@ private enum JsonPatchMember {
                                         remoteVersion = GetLongValue(valuePtr, valueSize, version.RemoteVersionPropertyName);
                                         if (remoteVersion != (version.RemoteVersion + 1)) {
                                             if (remoteVersion <= version.RemoteVersion) {
-                                                ThrowPatchException(patchStart, patchArrayPtr, patchArraySize, "Remote version mismatched.");
+                                                patchCount = -2;
```
Comment (joozek78) [inline src/Starcounter.XSON/Sessions/JSON-Patch/JsonPatch.cs]: that obviously shouldn't look like that, It was just quick-n-dirty way to make it return 203

File: src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs [hunk 3]
```diff
@@ -101,8 +102,15 @@ internal static class PuppetRestHandler {
 
                     if (patchCount == -1) { // -1 means that the patch was queued due to clientversion mismatch.
                         return new Response() {
+                            Resource = root,
                             StatusCode = 202,
-                            StatusDescription = "Patch enqueued until earlier versions have arrived"
+                            StatusDescription = "Patch enqueued until earlier versions have arrived. Last known version is " + root.ChangeLog.Version.RemoteVersion
+                        };
+                    } else if (patchCount == -2) {
```
Comment (tomalec) [inline src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs]: @joozek78  Could you give a note what does `-2` mean?

Comment (joozek78) [inline src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs]: I already merged this. But it's already explained [here](https://github.com/Starcounter/Starcounter/pull/3747/files#diff-3ba670736d50f315cde8d959007e3e42R540)

Comment (tomalec) [inline src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs]: I think you can add a comment after the merge.

It's explained in different file, for me it was very confusing to get what negative patchCount could mean, `-1` is explained https://github.com/Starcounter/Starcounter/pull/3747/files#diff-eb91747995c7619331005b1b26729025R103 but `-2` was a mystery.

Comment (joozek78) [inline src/Starcounter.Apps.JsonPatch/BuiltInRestHandlers/PuppetRestHandler.cs]: I've added it directly on develop

**Pull request merged**